src/providers/ipa/ipa_session.c

```diff
@@ -0,0 +1,834 @@
+                                                         state->gid);
+
+    state->pd->pam_status = (ret == EOK) ? PAM_SUCCESS : PAM_SESSION_ERR;
+
+done:
+    tevent_req_done(req);
+}
+
+errno_t
+ipa_pam_session_handler_recv(TALLOC_CTX *mem_ctx,
+                             struct tevent_req *req,
+                             struct pam_data **_data)
+{
+    struct ipa_pam_session_handler_state *state = NULL;
+
+    state = tevent_req_data(req, struct ipa_pam_session_handler_state);
+
+    TEVENT_REQ_RETURN_ON_ERROR(req);
+
+    *_data = talloc_steal(mem_ctx, state->pd);
+
+    return EOK;
+}
+
+static errno_t
+ipa_pam_session_handler_get_deskprofile_user_info(TALLOC_CTX *mem_ctx,
+                                                  struct sss_domain_info *domain,
+                                                  const char *username,
+                                                  char **_shortname,
+                                                  char **_domain,
+                                                  char **_user_dir,
+                                                  char **_domain_dir,
+                                                  uid_t *_uid,
+                                                  gid_t *_gid)
+{
+    TALLOC_CTX *tmp_ctx;
+    struct ldb_result *res = NULL;
+    char *shortname;
+    char *domain_name;
+    char *user_dir;
+    char *domain_dir;
+    uid_t uid;
+    gid_t gid;
+    errno_t ret;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    ret = sss_parse_internal_fqname(tmp_ctx, username,
+                                    &shortname, &domain_name);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_TRACE_FUNC, "Failed to parse \"%s\" [%d]: %s\n",
+              username, ret, sss_strerror(ret));
+        goto done;
+    }
+
+    domain_dir = talloc_asprintf(tmp_ctx,
+                                 IPA_DESKPROFILE_RULES_USER_DIR"/%s",
+                                 domain_name);
+    if (domain_dir == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc_asprintf() failed!\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    user_dir = talloc_asprintf(tmp_ctx, "%s/%s", domain_dir, shortname);
+    if (user_dir == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc_asprintf() failed!\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = sysdb_getpwnam(tmp_ctx, domain, username, &res);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "sysdb_getpwnam() failed [%d]: %s\n",
+              ret, sss_strerror(ret));
+        goto done;
+    }
+
+    if (res->count != 1) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "sysdb_getpwnam() got more users than expected. "
+              "Expected [%d], got [%d]\n", 1, res->count);
+        ret = EINVAL;
+        goto done;
+    }
+
+    uid = ldb_msg_find_attr_as_uint64(res->msgs[0], SYSDB_UIDNUM, 0);
+    gid = ldb_msg_find_attr_as_uint64(res->msgs[0], SYSDB_GIDNUM, 0);
```

Comment:
Since IPA doesn't handle root users or groups, I think we should be on the safe side and return an error with missing or zero UID or GID.